### DISCUSSION TITLE
Joystick with Create 3 Tele-operation

### DISCUSSION THREAD

**Discussion by Tokolosh4** [iRobotEducation/create3_docs#129] [Questions]
While the keyboard teleop works fine for my robot,  I am having trouble with installing an Xbox-360 controller.
This command does nothing:  `ros2 launch create3_teleop teleop_joystick_launch.py
`
I am following the instructions given by Justin here: [https://github.com/iRobotEducation/create3_docs/discussions/99](url)

I seem to have something wrong with my "joystick node".
Trying to disentangle this, I have followed instructions in  http://wiki.ros.org/joy/Tutorials/ConfiguringALinuxJoystick
I have configured the joystick as directed, can see its output change with:  `jstest /dev/input/js0`
However, starting the Joy Node, the command  `roscore`  gives an error:

> ImportError: cannot import name 'Log' from 'rosgraph_msgs.msg' (/opt/ros/galactic/lib/python3.8/site-packages/rosgraph_msgs/msg/__init__.py)

and the joy node does not start.

Suggestions for me please?
Thanks!

**Comment by justinIRBT** [comment]
roscore is a ROS 1 command.  The instructions you were looking at were for ROS 1.  Sounds like your joystick is working from your jstest results.

Can you show the output of running the launch file you tried earlier
ros2 launch create3_teleop teleop_joystick_launch.py

Also, someone had an issue where they didn’t press an enable button https://github.com/iRobotEducation/create3_docs/discussions/99#discussioncomment-3088024

**Reply by Tokolosh4** [reply]
Thanks for your speedy response.
Here is my output from the launch file:

> ubuntu@rpi-create3:~$ ros2 launch create3_teleop teleop_joystick_launch.py
[INFO] [launch]: All log files can be found below /home/ubuntu/.ros/log/2022-08-04-13-23-12-461108-rpi-create3-3829
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [joy_node-1]: process started with pid [3831]
[INFO] [teleop_node-2]: process started with pid [3833]
[teleop_node-2] [INFO] [1659619393.157198352] [TeleopTwistJoy]: Teleop enable button 2.
[teleop_node-2] [INFO] [1659619393.157393628] [TeleopTwistJoy]: Turbo on button 5.
[teleop_node-2] [INFO] [1659619393.157422887] [TeleopTwistJoy]: Linear axis x on 1 at scale 0.700000.
[teleop_node-2] [INFO] [1659619393.157455738] [TeleopTwistJoy]: Turbo for linear axis x is scale 1.500000.
[teleop_node-2] [INFO] [1659619393.157478275] [TeleopTwistJoy]: Angular axis yaw on 0 at scale 0.400000.
[teleop_node-2] [INFO] [1659619393.157499145] [TeleopTwistJoy]: Turbo for angular axis yaw is scale 1.000000.

It's not clear to me what an "enable button" is.  Perhaps I'm not understanding how to use the controller.  Are there instructions on what the various buttons on the controller do and how I would change them?
And: which button is which?  Are they the same as those indicated when I run jtest?

Thanks,

**Reply by shamlian** [reply]
Per
> [teleop_node-2] [INFO] [1659619393.157198352] [TeleopTwistJoy]: Teleop enable button 2.

You need to figure out which button is "button 2." I would recommend using `jstest` for that. You would need to hold that button down while operating the axes mapped to linear and angular motion, which appear to be "1" and "0" -- similarly, use `jstest` to find them. I can't tell you which buttons or axes are which because every joystick is different.

The documentation for teleop_twist_joy can be found [here](https://index.ros.org/p/teleop_twist_joy/github-ros2-teleop_twist_joy/).

**Reply by Tokolosh4** [reply]
Thanks, particularly for the direct pointer to the Ros2 documentation on this.
I continue to have the same issue -- I've identified the buttons but the Create3 robot still doesn't respond.
I guess I'll limp along using just the keyboard until I've learned more about Ros.

**Reply by brianabouchard** [reply]
Hi @Tokolosh4! I'm curious to know a bit more about your setup. I was having some issues with joy teleop as well. I'm curious if you have set up a namespace for your Create 3 robot? Also, what are you running ROS2 on? Is it a VM or are you using a RPi?

**Reply by Tokolosh4** [reply]
Hi Briana. I’m very early in my exploration of robotics. I have only a single robot so I haven’t set up a namespace. I’m running an RPi attached to the Create3, and have a Linux (Ubuntu) laptop to talk to it, do the Lidar/Slam mapping. (I was having a lot of trouble with both a VM and the Linux dual-boot on my older Apple laptop, so gave up on that direction.)

**Reply by brianabouchard** [reply]
Sounds awesome! I am new to ROS2 as of this summer, so I totally understand being early in the journey. It sounds like you have a perfect setup. I'll replicate it on my end, though I only have a PS3 controller rather than xbox. One more question for you...how is the xbox controller connected? Do you have it connected via USB cable to the RPi or is it connected to your laptop via USB or is it connected to one or the other via Bluetooth (I don't know the options with xbox)?

I've been messing around with the teleop joy ROS2 package (the one create3_teleop builds on). I'm hoping to eventually suggest some modifications to the create3_teleop package to make troubleshooting a little easier and have an option to easily disable the pesky enable button. All a work in progress, but I'd love to get to the root of your issue so I can work that into the suggestions.

**Reply by Tokolosh4** [reply]
Hi Briana, 
I bought an inexpensive Xbox-compatible controller and I’m hypothesising it may have some limitations that mess up its working with ROS2. It came with a USB dongle that I’ve connected to the RPI, which then sees it as a USB-connected device, not Bluetooth. It does see the button presses when I use jstest.  My thinking has been that I will want it to control the robot even when I’m not using the laptop. But your thoughts and suggestions for connections would be welcome.

**Comment by jerrynsr** [comment]
You  can try all the support joystick configurations.. go through one at a time.

There's  xbox,  ps3,  ps3-holonomic, atk3,  and xd3

Also to move the robot you need to move the joystick and press  the forward button

For a quick test  you could push each button and trigger while pointing the joystick in the "up" position.  
I think most joysticks will work with the predefined profile.  I have an 8bitDO SN30 Pro+   and it works well.

I use the atk3 joystick profile:
ros2 launch create3_teleop teleop_joystick_launch.py joy_config:=atk3

**Comment by Tokolosh4** [comment]
Thanks for the suggestions!
I have this now working with the xbox profile when the controller dongle is connected to my laptop.
It still doesn't work when connected to the RPi ....

**Reply by brianabouchard** [reply]
That's big progress, awesome news! Odd that it isn't working on the RPi, though. I reproduced your setup with a ps3 controller instead of the xbox one. Initially, it worked for a moment, then stopped working for some time and worked again consistently. The only thing I did that changed it from not working to working was restarted the RPi with the controller unplugged and then plugged the controller back in once it was up and running. However, I suspect that is not the issue in your case, and this was likely just a poor wiring connection on my end, but in the unlikely event that it helps. 

One thing that would be helpful to see is what the joy topic is publishing. To do this, open a second terminal window. In the first window run your normal teleop twist command. Once that is running, in the second window, try `ros2 topic echo /joy`. Similar to how jstest showed you with regards to the RPi seeing the controller, this should show you the same thing but for the ROS topic `joy`. Let us know if the topic is publishing button presses, etc. 

Also, if you are connecting to your RPi via SSH rather than a screen and keyboard, you can use `screen` to generate new terminal windows and the control-A button combination to switch between windows.

**Reply by jerrynsr** [reply]
When testing with the PI what did you use to configure the bluetooth connection ?    When I want the joystick to connect directly to the bot I use hcitool.    If you haven't used hcitool you should search for it on the net and check it out.

**Reply by brianabouchard** [reply]
I believe @Tokolosh4 is doing all of this through a USB connection for the joystick rather than bluetooth, which is the same thing I've done on my end.

**Comment by Tokolosh4** [comment]
Hi Briana, 

I bought an inexpensive Xbox-compatible controller and I’m hypothesising it may have some limitations that mess up its working with ROS2. It came with a USB dongle that I’ve connected to the RPI, which then sees it as a USB-connected device, not Bluetooth. My thinking has been that I will want it to control the robot even when I’m not using the laptop. But your thoughts and suggestions for connections would be welcome.

The RPi does see the button presses when I use jstest. I’ve not understood why Create3 has adapted the ROS2 teleop joystick drivers rather than simply pointing us at ROS and the instructions there. Doing so would have saved me going down a rabbit-hole. But I suspect there’s a good reason. 

David
